Try: Site Logo placeholder tweaks

jasmussen opened the pull request:
## Description

Fixes #35096. This improves the setup state of the Site Logo in a few ways. Here's what it looks like in trunk, square and basic:

<img width="823" alt="Screenshot 2021-10-06 at 12 26 13" src="https://user-images.githubusercontent.com/1204802/136198178-a8e5d19a-6d09-4e8d-8ea5-afc411e40d61.png">

Here's what the PR does:

1. It lets the setup state inherit the border radius set on the block itself. Currently by the block style variation ("rounded"), but theoretically by a separate border radius control in the future.
2. It makes the resting state of the block look like a dashed outline version of the Site Logo icon itself.

![site logo](https://user-images.githubusercontent.com/1204802/136359230-e4b806fc-cd10-400a-9f3e-e4fd6254b5d8.gif)

At the moment, you can resize from the default 120px dimensions by adding an image, resizing, then removing the image again, as shown in the GIF above. It's a bit of a roundabout way, but the placeholder is now smart enough to account for that:

<img width="823" alt="Screenshot 2021-10-07 at 11 27 02" src="https://user-images.githubusercontent.com/1204802/136359383-a867dc8b-7540-42aa-90d0-a4a35685f7bd.png">
<img width="790" alt="Screenshot 2021-10-07 at 11 27 08" src="https://user-images.githubusercontent.com/1204802/136359388-4224bf8f-75ac-48cf-92be-a91760936cbe.png">

## How has this been tested?

Insert a site logo, whether on its own or inside a navigation block. Test resizing it, test setting the rounded style, try adding a logo in both styles.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

jasmussen committed:
Initial work.

jasmussen committed:
Add illustration.

jasmussen committed:
Add animation.

jasmussen committed:
Border radius.

gwwar commented:
Nice, this feels pretty close to having all needed functionality. One thing I noticed, was that we need a different area to show any error messages. The placeholder here will be too small:

Edit: okay this is a problem in trunk too :D 

| branch | trunk |
|-----|-----|
| <img width="502" alt="CleanShot 2021-10-06 at 09 35 35@2x" src="https://user-images.githubusercontent.com/1270189/136246208-55102de2-c7b9-4255-9d5e-f31bd28a7764.png"> | <img width="495" alt="CleanShot 2021-10-06 at 09 39 50@2x" src="https://user-images.githubusercontent.com/1270189/136246948-8cb8dd22-3007-4f08-ba30-0cef3b196a6a.png"> |

#### After: using snackbar

https://user-images.githubusercontent.com/1270189/136281186-cc20e0c2-c9c7-4497-a92b-cfa485a873e6.mp4

gwwar commented:
I think this is our first usage of the non-scaling-stroke vector effect. I'm noting the browser support here. We dropped IE support, so I think we should be okay here.

<img width="1295" alt="CleanShot 2021-10-06 at 09 48 13@2x" src="https://user-images.githubusercontent.com/1270189/136248254-918bd567-bdf7-4bec-b872-f59c6d9141a9.png">

gwwar commented on packages/block-library/src/site-logo/editor.scss:
```diff
@@ -77,9 +80,80 @@
+			height: $grid-unit-60;
+			border-radius: 50%;
+			position: relative;
+			z-index: 1;
```
We can avoid z-index if we use natural stacking order. Might need another prop to note the placement of children.

gwwar commented on packages/block-library/src/site-logo/editor.scss:
```diff
@@ -77,9 +80,80 @@
 
-		// Hide drag and drop text.
+		// Hide items.
+		.components-placeholder__label,
```
Similarly we can pass through a prop to not render the label, and we can remove this. Edit: maybe allow label to be undefined? Placeholder is a base component so we might want to do this as a follow up PR

gwwar committed:
Site Logo: use custom media logo button

gwwar committed:
Site Logo: pass through custom placeholder and remove z-index rules 🔪

gwwar committed:
Site Logo: add a tooltip on hover

gwwar committed:
Site Logo: use snackbar for errors

gwwar commented:
@jasmussen I pushed up changes to add the upload icon, hover tooltip, moved the error notice to the snackbar, and provided some passthrough options to the MediaPlaceholder.

The tooltip interaction doesn't feel quite right to me, so I suspect that can use some further work/thinking. Feel free to modify anything that you think needs some further changes. 👍 

https://user-images.githubusercontent.com/1270189/136282036-67a1dfb3-1b7a-4f74-a181-eb8a81764514.mp4

jasmussen commented:
Thanks so much for doing all this work! It's a master class for me. 

I agree on the tooltip, but thanks for adding it. I'm going to try today and shift things in this direction:

> Make the upload button appear only after the block has first been selected. Probably worth experimenting with, as it might scale better to arbitrary logo sizes, but also good to make sure that the resize handles would still work well in this case.

That way, the tooltip can just be a tooltip that happens as usual when you hover the upload button, rather than the placeholder itself.

jasmussen committed:
Tweak the hover/select mechanism, polish.

jasmussen commented:
Such nice work, thanks so much for all your help. I managed to push a few tweaks getting it to this state:
![site logo](https://user-images.githubusercontent.com/1204802/136353859-71f0606e-b387-4f32-a790-a5f51efa7d7f.gif)

- The upload/media library button now appears after you select the block.
- The tooltip is now attached to that button, not the placeholder.

I feel like the behavior is now pretty solid. I'm going to rebase to fix a merge conflict, fix a couple of small bugs I discovered with resizing, and then mark this one as ready for review.

jasmussen committed:
Provide size smarts, and fix a radius issue.

jasmussen commented:
I rebased, the branch, fixed an issue with an inherited border radius, and added some placeholder sizing smarts:
![site logo](https://user-images.githubusercontent.com/1204802/136359230-e4b806fc-cd10-400a-9f3e-e4fd6254b5d8.gif)

At the moment, you can resize from the default 120px dimensions by adding an image, resizing, then removing the image again, as shown in the GIF above. It's a bit of a roundabout way, but the placeholder is now smart enough to account for that:

<img width="823" alt="Screenshot 2021-10-07 at 11 27 02" src="https://user-images.githubusercontent.com/1204802/136359383-a867dc8b-7540-42aa-90d0-a4a35685f7bd.png">
<img width="790" alt="Screenshot 2021-10-07 at 11 27 08" src="https://user-images.githubusercontent.com/1204802/136359388-4224bf8f-75ac-48cf-92be-a91760936cbe.png">

Separate to this PR it might be nice to introduce the resize component to the placeholder as well, so that it can just be sized right there.

All of this is also to say: this now feels solid to me, and I'm marking this one ready for review.

jasmussen committed:
Add comment to currentColor property.

gwwar commented on packages/block-library/src/site-logo/edit.js:
```diff
@@ -428,23 +427,52 @@ export default function LogoEdit( {
 		className: classes,
 	} );
+
+	const label = __( 'Add a site logo' );
```
Did we want to give a hint about drag&drop or clicking for the media library?

jasmussen replied:
Per a newer comment on the ticket, I thought it worth it to start small. But it reminds me, I forgot to probably test the drag and drop 🙈 I need to give that a quick spin. It's probably fine.

jasmussen replied:
Thanks so much again for all the help. I took the drag and drop for a spin and pushed a tiny tweak, otherwise it's looking good!

gwwar reviewed (approved):
Nice updates @jasmussen This tests nicely for me! I left a note about the tooltip copy, but it's not blocking. I agree on adding a follow up PR to allow resizing without an image set, it feels pretty natural to have that with a placeholder wireframe available.

We'll want a +1 review for this PR since I also pushed changes here. 

One thing for folks to look over are the new `mediaLibraryButton` and `placeholder` component passthrough props that I added to the block-editor/media-placeholder. Site Logo needed some more drastic component changes, but we can perhaps later refactor this into it's own component and pull out some of the upload checks into hooks.

jasmussen committed:
Add blinking fix.

jasmussen commented:
I pushed a small fix to a blinking behavior that happened when you drag/dropped an image onto the icon itself, so now it works as intended:

![dragondrop](https://user-images.githubusercontent.com/1204802/136432471-e21e9aae-40cb-4f4c-8f57-a36f60ffdc03.gif)

[That fix is also part of this separate PR that overhauls the drag and drop animations](https://github.com/WordPress/gutenberg/pull/33132/files#diff-efdedbd99be1e19d0b1e16c0df0ff62124b616aa6ec997c8f5c3408ac631785aR44).

Mamaduka reviewed (approved):
Approving this since I think enhancement looks good, and follow-up PRs can address my feedback.

I noticed that if I change the size and then reset the Site Logo, the placeholder shrinks:

https://user-images.githubusercontent.com/240569/136533567-3cb9a988-5416-49aa-988d-91275ec91b49.mp4

**New `placeholder` prop:** I think it can accept a component instead of a callback. Similar to how the `InnerBlocks` appender works. We're already passing all the props we might need to `MediaPlaceholder.`

We also need to document these two new props.

jasmussen commented:
I [ran into the same size issue](https://github.com/WordPress/gutenberg/pull/35397#issuecomment-937622851), and I think the solution is to allow the resize handles even in the setup state. The thing is, this particular block comes with a default size (120x120), which is applied if the block isn't manually resized. As soon as the block is manually resized, no min-width/heights are applied, which in the setup state means it collapses to the smallest button size.

However it makes sense for themes to be able to predefine dimensions of the site logo block, to fit theme metrics, and this would further be enhanced by being reflected in the setup state.

Thank you for the review! I'm going to land this one for now!
